### PR TITLE
SNOW-1825607 Initial OCSP deprecation plan steps

### PR DESCRIPTION
**Pull request by sfc-gh-ext-simba-jy**
### Description
Please explain the changes you made here.
- [x] FAIL_OPEN log changed (message + log level)?
- [x] FAIL_OPEN passing request along if responder URL is missing?
- [x] FAIL_CLOSED - no NPE if responder URL is missing? Please check the message below about this.
Updated: the driver will log the debug message and return callback with null if the OCSP status is FAIL_OPEN. If the status is FAIL_CLOSED, the driver will log the error message and return callback with the error.
- [x] Retries and timeouts confirmed (both to responder and cache server)?
- [x] `disableOCSPChecks` introduced?
  - [x] Deprecated `insecureMode`?
  - [x] `disableOCSPChecks` has precedence?
  
### Checklist
- [ ] Format code according to the existing code style (run `npm run lint:check -- CHANGED_FILES` and fix problems in changed code)
- [ ] Create tests which fail without the change (if possible)
- [ ] Make all tests (unit and integration) pass (`npm run test:unit` and `npm run test:integration`)
- [ ] Extend the README / documentation and ensure is properly displayed (if necessary)
- [ ] Provide JIRA issue id (if possible) or GitHub issue id in commit message

### CODE REVIEW

**Comment by sfc-gh-ext-simba-jy** [comment]
The node.js driver does not have the case that the responder URL does not exist according to the code: https://github.com/snowflakedb/snowflake-connector-nodejs/blob/bada542bbb8b46026a903d45b622562e65348e08/lib/agent/check.js#L176. As we have the default OCSP_RESPONSE_CACHE_SERVER_URL, this will not throw the error like JDBC.

File: lib/agent/socket_util.js [hunk 1]
```diff
@@ -120,7 +115,7 @@ exports.secureSocket = function (socket, host, agent, mock) {
 function isOcspValidationDisabled(host) {
   // ocsp is disabled if insecure-connect is enabled, or if we've disabled ocsp
   // for non-snowflake endpoints and the host is a non-snowflake endpoint
-  return GlobalConfig.isInsecureConnect() ||
+  return GlobalConfig.isOCSPChecksDisabled() ||
```
Comment (sfc-gh-pfus) [inline lib/agent/socket_util.js:118]: Does this cover also a case when someone sets `insecureMode`?

File: lib/agent/socket_util.js [hunk 2]
```diff
@@ -158,7 +153,7 @@ function canEarlyExitForOCSP(errors) {
       const err = errors[errorIndex];
       if (err && !isValidOCSPError(err)) {
         // any of the errors is NOT good/revoked/unknown
-        Logger.getInstance().warn(ocspFailOpenWarning + err);
+        Logger.getInstance().debug('OCSP responder didn\'t respond correctly. Assuming certificate is not revoked');
```
Comment (sfc-gh-pfus) [inline lib/agent/socket_util.js:156]: Can we leave details as it was before?

File: lib/core.js [hunk 3]
```diff
@@ -196,14 +196,18 @@ function Core(options) {
           Logger.getInstance().info('Configuring logger with level: %s, filePath: %s, additionalLogToConsole: %s', logLevel, logFilePath, additionalLogToConsole);
         }
 
-        const insecureConnect = options.insecureConnect;
-        if (Util.exists(insecureConnect)) {
+        if (Util.exists(options.insecureConnect)) {
+          Logger.getInstance().warn('Warning! The option insecureConnect will be deprecated. Please use the disableOCSPChecks.');
```
Comment (sfc-gh-pfus) [inline lib/core.js:200]: Actually it is deprecated, not will be :)

File: lib/core.js [hunk 4]
```diff
@@ -196,14 +196,18 @@ function Core(options) {
...
+        }
+
+        const disableOCSPChecks = options.insecureConnect || options.disableOCSPChecks;
```
Comment (sfc-gh-pfus) [inline lib/core.js:203]: Are this values nullable? If yes and both are not null, `disableOCSPChecks` should have precedence, not the alternative.

File: test/unit/ocsp/test_unit_ocsp_mode.js [hunk 5]
```diff
@@ -9,13 +9,13 @@ const assert = require('assert');
 describe('OCSP mode', function () {
   it('getOcspMode', function (done) {
     // insecure mode
-    GlobalConfig.setInsecureConnect(true);
+    GlobalConfig.setDisableOCSPChecks(true);
```
Comment (sfc-gh-pfus) [inline test/unit/ocsp/test_unit_ocsp_mode.js:12]: Can we have also a test confirming that `insecureMode` still works?

File: test/unit/snowflake_config_test.js [hunk 6]
```diff
@@ -15,7 +15,7 @@ describe('Snowflake Configure Tests', function () {
   before(function () {
     originalConfig = {
       logLevel: Logger.getInstance().getLevelTag(),
-      insecureConnect: GlobalConfig.isInsecureConnect(),
+      disableOCSPChecks: GlobalConfig.isOCSPChecksDisabled(),
```
Comment (sfc-gh-pfus) [inline test/unit/snowflake_config_test.js:18]: Same here + test for setting both values at once.

**Comment by sfc-gh-pfus** [comment]
> It seems like the node.js driver does not have the use case that the responder URL does not exist according to the code:
> 
> https://github.com/snowflakedb/snowflake-connector-nodejs/blob/bada542bbb8b46026a903d45b622562e65348e08/lib/agent/check.js#L176
> 
> . As we have the default OCSP_RESPONSE_CACHE_SERVER_URL and there will be no error for the responder URL, this will not throw the error like JDBC.

What you mentioned is not a responder URL - it is a cache server URL. Responder URL is provided in each certificate.

**Comment by sfc-gh-dszmolka** [comment]
> It seems like the node.js driver does not have the use case that the responder URL does not exist according to the code:
> 
> https://github.com/snowflakedb/snowflake-connector-nodejs/blob/bada542bbb8b46026a903d45b622562e65348e08/lib/agent/check.js#L176
> 
> . As we have the default OCSP_RESPONSE_CACHE_SERVER_URL and there will be no error for the responder URL, this will not throw the error like JDBC.


OCSP Responder url != OCSP Cache URL.  The (default) OCSP Cache URL is always the same, by default, `ocsp.snowflakecomputing.com` (gracefully ignoring privatelink situations here)

However OCSP Responder URL is not from Snowflake. It is always hardcoded into the particular certificate, which is under verification. See this random cert as an example: https://crt.sh/?sha256=9c3f2fd11c57d7c649ad5a0932c0f0d29756f6a0a1c74c43e1e89a62d64cd320
```
            Authority Information Access: 
                CA Issuers - URI:http://i.pki.goog/r4.crt
```

It _does not_ have OCSP endpoint. The driver today will complain if it encounters such certs which doesn't have the OCSP endpoint, because the driver expects the cert to have it. See this issue what exactly happens: https://github.com/snowflakedb/snowflake-connector-nodejs/issues/932

As a comparison, here's a cert which _does_ have OCSP endpoint: https://crt.sh/?id=12092745633
Observe:
```
            Authority Information Access: 
>>>>            OCSP - URI:http://ocsp.r2m03.amazontrust.com
                CA Issuers - URI:http://crt.r2m03.amazontrust.com/r2m03.cer
```
(this section is missing from the first cert)

The goal here, is for the driver if encountering a cert _without_ the said OCSP endpoint, then it should just log the debug-level message and gracefully continue as if the verification was successful. 
Implementing this change will also fix 932.

File: lib/agent/socket_util.js (hunk 2 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/agent/socket_util.js:156]: Done.

File: lib/core.js (hunk 4 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/core.js:203]: Changed.

File: lib/agent/socket_util.js (hunk 1 above)
Comment (sfc-gh-ext-simba-jy) [reply lib/agent/socket_util.js:118]: Suppose you mean ` insecureConnect`, Yes. In the core.js, the driver will take both insecureConnect and disableOCSPchecks. The disableOCSPChecks option is prioritized.

File: test/unit/ocsp/test_unit_ocsp_mode.js (hunk 5 above)
Comment (sfc-gh-ext-simba-jy) [reply test/unit/ocsp/test_unit_ocsp_mode.js:12]: I made testing for insecureConnect. Thank you.

File: test/unit/snowflake_config_test.js (hunk 6 above)
Comment (sfc-gh-ext-simba-jy) [reply test/unit/snowflake_config_test.js:18]: Done.

**Review by sfc-gh-pfus — APPROVED** [review]

File: index.d.ts [hunk 7]
```diff
@@ -219,10 +219,16 @@ declare module 'snowflake-sdk' {
         additionalLogToConsole?: boolean | null;
 
         /**
-         * Check the ocsp checking is off.
+         * @deprecated
+         * This option will be deprecated. Use disableOCSPChecks
```
Comment (sfc-gh-pfus) [inline index.d.ts:223]: `is deprecated` instead of `will be`

**Review by sfc-gh-pfus — APPROVED** [review]

File: index.d.ts (hunk 7 above)
Comment (sfc-gh-ext-simba-jy) [reply index.d.ts]: I fixed it. Thank you.

**Comment by sfc-gh-pmotacki** [comment]
Hi @sfc-gh-ext-simba-jy.  Regarding the plan for BCR release in the next version, we have an option to remove the old parameter at all, not only set it deprecated.

**Comment by sfc-gh-ext-simba-jy** [comment]
@sfc-gh-pmotacki, I removed the old option (insecureConnect). Please have a look.

**Review by sfc-gh-pmotacki — APPROVED** [review]
LGTM

**Pull request merged**